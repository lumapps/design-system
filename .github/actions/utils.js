const util = require('util');
const exec = util.promisify(require('child_process').exec);

/**
 * Get short SHA for git ref.
 */
async function getShortSHA(gitRef) {
    return exec(`git rev-parse --short ${gitRef}`)
        .then(({ stdout }) => stdout.trim());
}

/**
 * Generate StoryBook URL for the given git ref.
 */
async function getStoryBookURL(gitRef) {
    const { project_id } = require('../../configs/chromatic');
    const shortSHA = await getShortSHA(gitRef);
    return `https://${shortSHA}--${project_id}.chromatic.com/`;
}

module.exports = { getStoryBookURL, getShortSHA };
