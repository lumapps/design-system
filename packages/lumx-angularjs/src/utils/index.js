/* eslint-disable import/no-unassigned-import */

import './depth_service';
import './enter-keydown_directive';
import './event-scheduler_service';
import './focus-on-init_directive';
import './focus-trap_service';
import './stop-propagation_directive';
import './utils_service';
