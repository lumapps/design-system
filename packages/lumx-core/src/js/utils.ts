import classNames from 'classnames';
// eslint-disable-next-line import/no-extraneous-dependencies
import React from 'react';

import isBoolean from 'lodash/isBoolean';
import isEmpty from 'lodash/isEmpty';
import kebabCase from 'lodash/kebabCase';
import noop from 'lodash/noop';

/**
 * Enhance isEmpty method to also works with numbers.
 *
 * @param  value The value to check.
 * @return Whether the input value is empty or != 0.
 */
const _isEmpty = (value: any) => {
    if (typeof value === 'number') {
        return value === 0;
    }

    return isEmpty(value);
};

/**
 * Get the basic CSS class for the given type.
 *
 * @param  prefix The class name prefix for the generated CSS class.
 * @param  type   The type of CSS class we want to generate (e.g.: 'color', 'variant', ...).
 * @param  value  The value of the type of the CSS class (e.g.: 'primary', 'button', ...).
 * @return The basic CSS class.
 */
export function getBasicClass({
    prefix,
    type,
    value,
}: {
    prefix: string;
    type: string;
    value: string | number | boolean | undefined;
}): string {
    if (isBoolean(value)) {
        if (!value) {
            // False value should not return a class.
            return '';
        }
        const booleanPrefixes = ['has', 'is'];

        if (booleanPrefixes.some((booleanPrefix) => type.toString().startsWith(booleanPrefix))) {
            return `${prefix}--${kebabCase(type)}`;
        }

        return `${prefix}--is-${kebabCase(type)}`;
    }

    return `${prefix}--${kebabCase(type)}-${value}`;
}

/**
 * Return all basic LumX CSS classes which are available for every components.
 *
 * @see {@link /src/components/index.d.ts} for the possible values of each parameter.
 *
 * @param  prefix The class name prefix for the generated CSS class.
 * @param  props  All the other props you want to generate a class.
 *                The rule of thumb: the key is the name of the prop in the class, the value a string that will
 *                be used in the classname to represent the value of the given prop.
 * @return All LumX basic CSS classes.
 */
export function handleBasicClasses({ prefix, ...props }: { prefix: string; [prop: string]: any }): string {
    const otherClasses: any = {};
    if (!isEmpty(props)) {
        Object.keys(props).forEach((prop) => {
            otherClasses[getBasicClass({ prefix, type: prop, value: props[prop] })] = isBoolean(props[prop])
                ? props[prop]
                : !_isEmpty(props[prop]);
        });
    }

    return classNames(prefix, otherClasses);
}

declare type SwipeDirection = 'none' | 'up' | 'down' | 'left' | 'right';

/**
 * Detects swipe direction.
 * Credits: http://javascriptkit.com/javatutors/touchevents2.shtml.
 *
 * @deprecated use `detectHorizontalSwipe` instead if possible (better performance and does not block scroll)
 * @param  touchSurface          Element that will hold touch events.
 * @param  handleSwipe Callback function.
 * @return Function to remove listeners.
 */
export function detectSwipe(touchSurface: Element, handleSwipe: (direction: SwipeDirection) => void = noop) {
    let distX: number;
    let distY: number;
    let startX: number;
    let startY: number;
    let direction: SwipeDirection;
    // Required min distance traveled to be considered swipe.
    const threshold = 150;
    // Maximum distance allowed at the same time in perpendicular direction.
    const restraint = 100;
    // Maximum time allowed to travel that distance.
    const allowedTime = 300;
    let elapsedTime: number;
    let startTime: number;

    const onTouchStart = (evt: Event) => {
        const [touch] = Array.from((evt as TouchEvent).changedTouches);
        direction = 'none';
        // Const dist = 0;
        startX = touch.pageX;
        startY = touch.pageY;
        // Record time when finger first makes contact with surface.
        startTime = new Date().getTime();
        evt.preventDefault();
    };

    const onTouchMove = (evt: Event) => {
        // Prevent scrolling when inside DIV.
        evt.preventDefault();
    };

    const onTouchEnd = (evt: Event) => {
        const [touch] = Array.from((evt as TouchEvent).changedTouches);
        // Get horizontal dist traveled by finger while in contact with surface.
        distX = touch.pageX - startX;
        // Get vertical dist traveled by finger while in contact with surface.
        distY = touch.pageY - startY;
        // Get time elapsed.
        elapsedTime = new Date().getTime() - startTime;
        if (elapsedTime <= allowedTime) {
            // First condition for awipe met.
            if (Math.abs(distX) >= threshold && Math.abs(distY) <= restraint) {
                // 2nd condition for horizontal swipe met.
                // If dist traveled is negative, it indicates left swipe.
                direction = distX < 0 ? 'left' : 'right';
            } else if (Math.abs(distY) >= threshold && Math.abs(distX) <= restraint) {
                // 2nd condition for vertical swipe met.
                // If dist traveled is negative, it indicates up swipe.
                direction = distY < 0 ? 'up' : 'down';
            }
        }
        handleSwipe(direction);
        evt.preventDefault();
    };

    touchSurface.addEventListener('touchstart', onTouchStart, false);
    touchSurface.addEventListener('touchmove', onTouchMove, false);
    touchSurface.addEventListener('touchend', onTouchEnd, false);

    return () => {
        touchSurface.removeEventListener('touchstart', onTouchStart, false);
        touchSurface.removeEventListener('touchmove', onTouchMove, false);
        touchSurface.removeEventListener('touchend', onTouchEnd, false);
    };
}

/**
 * Checks whether or not the browser support passive events.
 * @see https://github.com/Modernizr/Modernizr/blob/6d56d814b9682843313b16060adb25a58d83a317/feature-detects/dom/passiveeventlisteners.js
 */
function isPassiveEventAvailable() {
    let supportsPassiveOption = false;
    try {
        const opts = Object.defineProperty({}, 'passive', {
            get() {
                supportsPassiveOption = true;
            },
        });
        window.addEventListener('testPassiveEventSupport', noop, opts);
        window.removeEventListener('testPassiveEventSupport', noop, opts);
    } catch (e) {
        // ignored
    }
    return supportsPassiveOption;
}

/**
 * Detects horizontal swipe direction without blocking the browser scroll using passive event.
 * @see http://javascriptkit.com/javatutors/touchevents2.shtml
 * @see https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md
 */
export function detectHorizontalSwipe(touchSurface: Element, handleSwipe: (direction: 'right' | 'left') => void) {
    let startX: number;
    let startY: number;
    // Required min distance traveled to be considered swipe.
    const threshold = 150;
    // Maximum distance allowed at the same time in perpendicular direction.
    const restraint = 150;
    // Maximum time allowed to travel that distance.
    const allowedTime = 300;
    let elapsedTime: number;
    let startTime: number;
    let finished: boolean;

    const onTouchStart = (evt: Event) => {
        const [touch] = Array.from((evt as TouchEvent).changedTouches);
        startX = touch.pageX;
        startY = touch.pageY;
        // Record time when finger first makes contact with surface.
        startTime = new Date().getTime();
        finished = false;
    };

    const onTouchMove = (evt: Event) => {
        if (finished) {
            return;
        }

        elapsedTime = new Date().getTime() - startTime;
        if (elapsedTime > allowedTime) {
            // Touch swipe too long to be considered.
            return;
        }

        const [touch] = Array.from((evt as TouchEvent).changedTouches);
        // Get horizontal dist traveled by finger while in contact with surface.
        const distX = touch.pageX - startX;
        // Get vertical dist traveled by finger while in contact with surface.
        const distY = touch.pageY - startY;

        if (!(Math.abs(distX) >= threshold && Math.abs(distY) <= restraint)) {
            // Swipe is not horizontal.
            return;
        }
        // Swipe direction.
        const direction = distX < 0 ? 'left' : 'right';

        handleSwipe(direction);
        finished = true;
    };

    // Activate passive event if possible for better scrolling performance.
    const eventOptions: any = isPassiveEventAvailable() ? { passive: true } : false;
    touchSurface.addEventListener('touchstart', onTouchStart, eventOptions);
    touchSurface.addEventListener('touchmove', onTouchMove, eventOptions);

    return () => {
        touchSurface.removeEventListener('touchstart', onTouchStart, eventOptions);
        touchSurface.removeEventListener('touchmove', onTouchMove, eventOptions);
    };
}

type KeyboardEventHandler<E extends KeyboardEvent | React.KeyboardEvent> = (event: E) => void;

/**
 * Make sure the pressed key is the enter key before calling the callback.
 *
 * @param  handler The handler to call on enter/return press.
 * @return The decorated function.
 */
export function onEnterPressed<E extends KeyboardEvent | React.KeyboardEvent>(
    handler: KeyboardEventHandler<E>,
): KeyboardEventHandler<E> {
    return (evt) => {
        if (evt.key !== 'Enter') {
            return;
        }
        handler(evt);
    };
}

/**
 * Make sure the pressed key is the escape key before calling the callback.
 *
 * @param  handler The handler to call on enter/return press.
 * @return The decorated function.
 */
export function onEscapePressed<E extends KeyboardEvent | React.KeyboardEvent>(
    handler: KeyboardEventHandler<E>,
): KeyboardEventHandler<E> {
    return (evt) => {
        if (evt.key !== 'Escape') {
            return;
        }
        handler(evt);
    };
}
