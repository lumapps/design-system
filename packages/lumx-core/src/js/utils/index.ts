export * from './detectHorizontalSwipe';
export * from './detectSwipe';
export * from './getBasicClass';
export * from './handleBasicClasses';
export * from './isPassiveEventAvailable';
export * from './onButtonPressed';
export * from './onEnterPressed';
export * from './onEscapePressed';
