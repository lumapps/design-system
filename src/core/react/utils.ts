import React, { Children, ReactNode } from 'react';

import get from 'lodash/get';
import isBoolean from 'lodash/isBoolean';
import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';
import isString from 'lodash/isString';
import kebabCase from 'lodash/kebabCase';
import noop from 'lodash/noop';
import trimStart from 'lodash/trimStart';

import { CSS_PREFIX } from '../constants';

import { concat, dropRight, last, partition, reduce } from 'lodash';
import { COMPONENT_PREFIX } from './constants';

/////////////////////////////
//                         //
//    Private attributes   //
//                         //
/////////////////////////////

/**
 * The properties of a component to use to determine it's name.
 * In the order of preference.
 */
const NAME_PROPERTIES: string[] = ['displayName', 'name', 'type', 'type.name', '_reactInternalFiber.elementType.name'];

/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * Define a generic props types.
 */
interface IGenericProps {
    /**
     * Any prop (particularly any supported prop for a HTML element).
     * E.g. classNames, onClick, disabled, ...
     */
    // tslint:disable-next-line: no-any
    [propName: string]: any;
}

/**
 * Defines a generic component type.
 */
// tslint:disable-next-line: no-any
type ComponentType = ReactNode | React.FC<any> | React.PureComponent<any, any> | React.Component<any, any>;
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

interface IChildrenManipulationParameters {
    child: ReactNode;
    children: ReactNode;
    childrenCount: number;
    index: number;
    props: IGenericProps;
}
type ChildTransformParameters = IChildrenManipulationParameters;
type ChildValidateParameters = IChildrenManipulationParameters;

/**
 * Defines the parameters of the pre/post validate callback of the `validateComponent` function below.
 */
interface IValidateParameters {
    children: ReactNode;
    childrenCount: number;
    props: IGenericProps;
}
type ValidateParameters = IValidateParameters;

/////////////////////////////
//                         //
//     Public functions    //
//                         //
/////////////////////////////

/**
 * Get the name of the root CSS class of a component based on its name.
 *
 * @param componentName The name of the component. This name should contains the component prefix and be
 *                               written in PascalCase.
 * @param subComponent Whether the current component is a sub component, if true, define the class according
 *                               to BEM standards.
 * @return The name of the root CSS class. This classname include the CSS classname prefix and is written in
 *                  lower-snake-case.
 */
function getRootClassName(componentName: string, subComponent?: boolean): string {
    const formattedClassName = `${CSS_PREFIX}-${kebabCase(trimStart(componentName, COMPONENT_PREFIX))}`;

    if (subComponent) {
        // See https://regex101.com/r/YjS1uI/3
        const lastPieceOfClassNameRegExp: RegExp = /^(.*)-(.+)$/gi;

        return formattedClassName.replace(lastPieceOfClassNameRegExp, '$1__$2');
    }

    return formattedClassName;
}

/**
 * Get the name of the given type.
 * The type can either be a string ('Button', 'span', ...) or a component whose name will be computed from its
 * `displayName`, its React internal name (`_reactInternalFiber.elementType.name`) or its `name`.
 *
 * @param type The type to get the name of.
 * @return The name of the type.
 */
function getTypeName(type: string | ComponentType): string | ComponentType {
    if (isString(type)) {
        return type;
    }

    for (const nameProperty of NAME_PROPERTIES) {
        const typeName: string | undefined = get(type, nameProperty);

        if (typeName !== undefined && isString(typeName) && !isEmpty(typeName)) {
            return typeName;
        }
    }

    return type;
}

/**
 * Check if a ReactElement is of the given type.
 *
 * @param el   The ReactElement we want to check the type of.
 * @param type The type we want to check if the ReactElement is of.
 * @return     If the ReactElement is of the given type or not.
 */
function isElementOfType(el: ReactNode, type: string | ComponentType): boolean {
    const typeName: string | ComponentType = getTypeName(type);

    if (!isString(typeName) || isEmpty(typeName)) {
        console.debug('Un-computable type', type, '\nResulted in', typeName);
        throw new Error(
            `The type you want to check is not valid. Waiting a JSX element, a component or a string (got \`${type!.toString()}\`)`,
        );
    }

    for (const nameProperty of NAME_PROPERTIES) {
        const elementTypeName: string | undefined = get(el, nameProperty);

        if (
            elementTypeName !== undefined &&
            isString(elementTypeName) &&
            !isEmpty(elementTypeName) &&
            elementTypeName === typeName
        ) {
            return true;
        }
    }
    return false;
}

/**
 * Check if a ReactElement is a text (i.e. either a pure text node or a <span>).
 * Simply check if the ReactElement is a string or if its type is 'span'.
 *
 * @param el The ReactElement to check if it's a text.
 * @return   If the ReactElement is a text or not.
 */
function isElementText(el: ReactNode): boolean {
    return isString(el);
}

/**
 * Unwrap the children contained in a fragment.
 *
 * @param children The children to unwrap (there should be only 1 child of React.Fragment type).
 * @return The unwrapped children (or the origin children if it wasn't a fragment).
 */
function unwrapFragment(children: ReactNode): ReactNode {
    let newChildren: ReactNode = children;

    const childrenCount: number = Children.count(children);
    if (childrenCount === 1) {
        const firstChild: ReactNode = Children.toArray(children)[0];

        if (get(firstChild, 'type') === React.Fragment) {
            newChildren = get(firstChild, 'props.children', []);
        }
    }

    return newChildren;
}

/**
 * Validate the component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param componentName    The name of the component being validated.
 * @param [allowedTypes]   The allowed types of children.
 * @param [maxChildren]    The maximum expected number of children.
 * @param [minChildren=0]  The minimum expected number of children.
 * @param [postValidate]   A function to run after all the transformation and validation
 *                                                        of the children and the props.
 *                                                        This function can return a string (the error message), a
 *                                                        boolean (`true` for a successful validation, `false` for a bad
 *                                                        validation which will lead to throw a basic error message) or
 *                                                        nothing if there is no special problem (i.e. a successful
 *                                                        validation).
 * @param [preValidate]    A function to run before global validation of the props and
 *                                                        any transformation or validation of the children.
 *                                                        This function can return a string (the error message), a
 *                                                        boolean (`true` for a successful validation, `false` for a bad
 *                                                        validation which will lead to throw a basic error message) or
 *                                                        nothing if there is no special problem (i.e. a successful
 *                                                        validation).
 * @param props            The props of the component (should contain a `children` prop).
 * @param [transformChild] A function to transform a child to something else.
 * @param [validateChild]  A function to check if a child is valid after that its type
 *                                                        has been validated (if `allowedTypes` is provided) and after
 *                                                        it has been transformed..
 *                                                        This function can return a string (the error message), a
 *                                                        boolean (`true` for a successful validation, `false` for a bad
 *                                                        validation which will lead to throw a basic error message) or
 *                                                        nothing if there is nothing special (i.e. a successful
 *                                                        validation).
 * @return                 The processed children of the component.
 */
function validateComponent(
    componentName: string,
    {
        allowedTypes = [],
        maxChildren,
        minChildren = 0,
        props,
        postValidate = noop,
        preValidate = noop,
        transformChild,
        validateChild,
    }: {
        allowedTypes?: Array<string | ComponentType>;
        maxChildren?: number;
        minChildren?: number;
        props: IGenericProps;
        postValidate?(params: ValidateParameters): string | boolean | void;
        preValidate?(params: ValidateParameters): string | boolean | void;
        transformChild?(params: ChildTransformParameters): ReactNode;
        validateChild?(params: ChildValidateParameters): string | boolean | void;
    },
): ReactNode | undefined {
    let newChildren: ReactNode = !isEmpty(props.children) ? unwrapFragment(props.children) : undefined;
    const childrenCount: number = Children.count(newChildren);

    const preValidation: string | boolean | void = preValidate({ children: newChildren, childrenCount, props });
    if (isString(preValidation) || (isBoolean(preValidation) && !preValidation)) {
        throw new Error(
            `Pre-validation of <${componentName}> failed${isString(preValidation) ? `: '${preValidation}'` : ''}.`,
        );
    }

    if (minChildren !== 0) {
        if (maxChildren === minChildren && childrenCount !== minChildren) {
            throw new Error(
                `<${componentName}> must have exactly ${minChildren} ${
                    minChildren === 1 ? 'child' : 'children'
                } (got ${childrenCount})!`,
            );
        }

        if (childrenCount < minChildren) {
            throw new Error(
                `<${componentName}> must have at least ${minChildren} ${
                    minChildren <= 1 ? 'child' : 'children'
                } (got ${childrenCount})!`,
            );
        }
    } else if (maxChildren === minChildren && childrenCount > minChildren) {
        throw new Error(`<${componentName}> shouldn't have any child (got ${childrenCount})!`);
    }

    if (maxChildren !== undefined && maxChildren !== minChildren && childrenCount > maxChildren) {
        throw new Error(
            `<${componentName}> cannot have more than ${maxChildren} ${
                maxChildren === 1 ? 'child' : 'children'
            } (got ${childrenCount})!`,
        );
    }

    if (isFunction(transformChild) || isFunction(validateChild) || !isEmpty(allowedTypes)) {
        const childrenFunctionName: string = isFunction(transformChild) ? 'map' : 'forEach';

        if (!isFunction(transformChild)) {
            transformChild = ({ child }: ChildTransformParameters): ReactNode => child;
        }
        validateChild = validateChild || noop;

        let index = -1;
        const transformedChildren: ReactNode = Children[childrenFunctionName](
            newChildren,
            (child: ReactNode): ReactNode => {
                index++;

                const newChild: ReactNode = transformChild!({
                    child,
                    children: newChildren,
                    childrenCount,
                    index,
                    props,
                });
                const childDisplayName: string = isString(newChild) ? `'${newChild}'` : `<${getTypeName(newChild)}>`;

                if (!isEmpty(allowedTypes)) {
                    const isOfOneAllowedType: boolean = allowedTypes.some(
                        (allowedType: string | ComponentType) =>
                            isEmpty(newChild) ||
                            (allowedType === 'text'
                                ? isElementText(newChild)
                                : isElementOfType(newChild, allowedType) || isEmpty(newChild)),
                    );

                    if (!isOfOneAllowedType) {
                        let allowedTypesString = '';
                        allowedTypes.forEach(
                            (allowedType: string | ComponentType, idx: number): void => {
                                if (!isEmpty(allowedTypesString)) {
                                    allowedTypesString += idx < allowedTypes.length - 1 ? ', ' : ' or ';
                                }

                                const typeName: string | ComponentType = getTypeName(allowedType);
                                allowedTypesString +=
                                    (isString(typeName) ? `${typeName === 'text' ? typeName : `<${typeName}>`}` : '') ||
                                    '<Unknown component type>';
                            },
                        );

                        console.debug('Non matching type', newChild, '\nResulted in', getTypeName(newChild));
                        throw new Error(
                            `You can only have ${allowedTypesString} children in <${componentName}> (got ${childDisplayName})!`,
                        );
                    }
                }

                const childValidation: string | boolean | void = validateChild!({
                    child: newChild,
                    children: newChildren,
                    childrenCount,
                    index,
                    props,
                });
                if (isString(childValidation) || (isBoolean(childValidation) && !childValidation)) {
                    throw new Error(
                        `Child validation of ${childDisplayName} in ${componentName} failed${
                            isString(childValidation) ? `: ${childValidation}` : ''
                        }'.`,
                    );
                }

                return newChild;
            },
        );

        if (childrenFunctionName === 'map') {
            newChildren = transformedChildren;
        }
    }

    const postValidation: string | boolean | void = postValidate({ children: newChildren, childrenCount, props });
    if (isString(postValidation) || (isBoolean(postValidation) && !postValidation)) {
        throw new Error(
            `Post-validation of ${componentName} failed${isString(postValidation) ? `: ${postValidation}` : ''}.`,
        );
    }

    return newChildren;
}

/**
 * Callback function type alias (use for readability)
 */
type Callback = () => void;

type Predicate<T> = (t: T) => boolean;

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

/**
 * Create a predicate function that checks if a ReactNode is a react element from the given component.
 *
 * @param  component React function component or the component name
 * @return predicate returning true if value is instance of the component
 */
const isComponent = <C>(component: React.FC<C> | string): Predicate<ReactNode> => (instance: ReactNode): boolean => {
    const instanceTypes = ['type', 'type.displayName', 'props.mdxType'];
    const componentName = typeof component === 'string' ? component : component.displayName;

    return (
        get(instance, '$$typeof') === Symbol.for('react.element') &&
        instanceTypes.some((type: string): boolean => get(instance, type) === componentName)
    );
};

/////////////////////////////

export {
    ChildTransformParameters,
    ChildValidateParameters,
    ComponentType,
    IGenericProps,
    Omit,
    ValidateParameters,
    getRootClassName,
    getTypeName,
    isElementOfType,
    isElementText,
    unwrapFragment,
    validateComponent,
    Callback,
    partitionMulti,
    Predicate,
    isComponent,
};
