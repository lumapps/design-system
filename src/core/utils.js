import classNames from 'classnames';

import isBoolean from 'lodash/isBoolean';
import isEmpty from 'lodash/isEmpty';
import kebabCase from 'lodash/kebabCase';
import { noop } from 'lodash/noop';

import { ENTER_KEY_CODE, ESCAPE_KEY_CODE } from './constants';

/**
 * Enhance isEmpty method to also works with numbers.
 *
 * @param  {any}     value The value to check
 * @return {boolean} Weither if the input value is empty or != 0.
 */
const _isEmpty = (value) => {
    if (typeof value === 'number') {
        return value === 0;
    }

    return isEmpty(value);
};

/////////////////////////////
//                         //
//     Public functions    //
//                         //
/////////////////////////////

/**
 * Get the basic CSS class for the given type.
 *
 * @param  {string}         prefix The class name prefix for the generated CSS class.
 * @param  {string}         type   The type of CSS class we want to generate (e.g.: 'color', 'variant', ...).
 * @param  {string|boolean} value  The value of the type of the CSS class (e.g.: 'primary', 'button', ...).
 * @return {string}         The basic CSS class.
 */
function getBasicClass({ prefix, type, value }) {
    if (isBoolean(value)) {
        if (!value) {
            // false value should not return a class
            return '';
        }
        const booleanPrefixes = ['has', 'is'];

        if (booleanPrefixes.some((booleanPrefix) => type.toString().startsWith(booleanPrefix))) {
            return `${prefix}--${kebabCase(type)}`;
        }

        return `${prefix}--is-${kebabCase(type)}`;
    }

    return `${prefix}--${kebabCase(type)}-${value}`;
}

/**
 * Return all basic LumX CSS classes which are available for every components.
 *
 * @see {@link /src/components/index.d.ts} for the possible values of each parameter.
 *
 * @param  {string} prefix The class name prefix for the generated CSS class.
 * @param  {Object} props  All the other props you want to generate a class.
 *                         The rule of thumb: the key is the name of the prop in the class, the value a string that will
 *                         be used in the classname to represent the value of the given prop.
 * @return {string} All LumX basic CSS classes.
 */
function handleBasicClasses({ prefix, ...props }) {
    const otherClasses = {};
    if (!isEmpty(props)) {
        Object.keys(props).forEach((prop) => {
            otherClasses[getBasicClass({ prefix, type: prop, value: props[prop] })] = isBoolean(props[prop])
                ? props[prop]
                : !_isEmpty(props[prop]);
        });
    }

    return classNames(prefix, otherClasses);
}

/**
 * Detects swipe direction.
 * Credits: http://javascriptkit.com/javatutors/touchevents2.shtml.
 *
 * @param  {Element}  el Element that will hold touch events.
 * @param  {Function} cb Callback function.
 * @return {Function} Function to remove listeners.
 */
function detectSwipe(el, cb = noop) {
    const touchsurface = el;
    let distX, distY, startX, startY, swipedir;
    // Required min distance traveled to be considered swipe.
    const threshold = 150;
    // Maximum distance allowed at the same time in perpendicular direction.
    const restraint = 100;
    // Maximum time allowed to travel that distance.
    const allowedTime = 300;
    let elapsedTime, startTime;
    const handleswipe = cb;

    const onTouchStart = (evt) => {
        const [touchobj] = evt.changedTouches;
        swipedir = 'none';
        // Const dist = 0;
        startX = touchobj.pageX;
        startY = touchobj.pageY;
        // Record time when finger first makes contact with surface.
        startTime = new Date().getTime();
        evt.preventDefault();
    };

    const onTouchMove = (evt) => {
        // Prevent scrolling when inside DIV.
        evt.preventDefault();
    };

    const onTouchEnd = (evt) => {
        const [touchobj] = evt.changedTouches;
        // Get horizontal dist traveled by finger while in contact with surface.
        distX = touchobj.pageX - startX;
        // Get vertical dist traveled by finger while in contact with surface.
        distY = touchobj.pageY - startY;
        // Get time elapsed.
        elapsedTime = new Date().getTime() - startTime;
        if (elapsedTime <= allowedTime) {
            // First condition for awipe met.
            if (Math.abs(distX) >= threshold && Math.abs(distY) <= restraint) {
                // 2nd condition for horizontal swipe met.
                // If dist traveled is negative, it indicates left swipe.
                swipedir = distX < 0 ? 'left' : 'right';
            } else if (Math.abs(distY) >= threshold && Math.abs(distX) <= restraint) {
                // 2nd condition for vertical swipe met.
                // If dist traveled is negative, it indicates up swipe.
                swipedir = distY < 0 ? 'up' : 'down';
            }
        }
        handleswipe(swipedir);
        evt.preventDefault();
    };

    touchsurface.addEventListener('touchstart', onTouchStart, false);
    touchsurface.addEventListener('touchmove', onTouchMove, false);
    touchsurface.addEventListener('touchend', onTouchEnd, false);

    return () => {
        touchsurface.removeEventListener('touchstart', onTouchStart, false);
        touchsurface.removeEventListener('touchmove', onTouchMove, false);
        touchsurface.removeEventListener('touchend', onTouchEnd, false);
    };
}

/**
 * Make sure the pressed key is the enter key before calling the callback.
 *
 * @param  {Function} cb The callback to call on enter/return press.
 * @return {Function} The decorated function.
 */
function onEnterPressed(cb) {
    return (evt) => {
        if (evt.keyCode !== ENTER_KEY_CODE) {
            return;
        }

        cb();
    };
}

/**
 * Make sure the pressed key is the escape key before calling the callback.
 *
 * @param  {Function} cb The callback to call on escape press.
 * @return {Function} The decorated function.
 */
function onEscapePressed(cb) {
    return (evt) => {
        if (evt.keyCode !== ESCAPE_KEY_CODE) {
            return;
        }

        cb();
    };
}

/////////////////////////////

export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed };
