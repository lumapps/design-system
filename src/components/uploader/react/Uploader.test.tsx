import React, { ReactElement } from 'react';

import { mount, shallow } from 'enzyme';

import { ICommonSetup, Wrapper, commonTestsSuite } from 'LumX/core/testing/utils.test';
import { getBasicClass } from 'LumX/core/utils';

import { CLASSNAME, DEFAULT_PROPS, Uploader, UploaderProps } from './Uploader';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<UploaderProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * [Enter the description of this wrapper].
     * [You should also probably change the name of the wrapper to something more meaningful].
     */
    wrapper: Wrapper;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  props  The props to use to override the default props of the component.
 * @param  [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return An object with the props, the component wrapper and some shortcut to some element inside of the component.
 */
const setup = (props: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    // @ts-ignore
    const wrapper: Wrapper = renderer(<Uploader {...props} />);

    return {
        props,
        wrapper,
    };
};

describe(`<${Uploader.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        // Here is an example of a basic rendering check, with snapshot.

        it('should render correctly', (): void => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        // Here are some examples of basic props check.

        it('should use default props', (): void => {
            const { wrapper } = setup();

            Object.keys(DEFAULT_PROPS).forEach((prop: string): void => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: DEFAULT_PROPS[prop] }),
                );
            });
        });
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Here is an example how to check a `onClick` event.

        const onClick: jest.Mock = jest.fn();

        beforeEach((): void => {
            onClick.mockClear();
        });

        it('should trigger `onClick` when clicked', (): void => {
            const { wrapper } = setup({ onClick }, false);

            wrapper.simulate('click');

            expect(onClick).toHaveBeenCalled();
        });
    });
    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
