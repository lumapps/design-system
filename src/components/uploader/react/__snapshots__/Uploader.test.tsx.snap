// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Uploader> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-uploader lumx-uploader--aspect-ratio-horizontal lumx-uploader--size-xl lumx-uploader--theme-light lumx-uploader--variant-square"
>
  <div
    className="lumx-uploader__background"
  />
  <div
    className="lumx-uploader__wrapper"
  />
</div>
`;
